File: app/src/main/java/id/rataapk/template/GitHubSettings.java
package id.rataapk.template;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

/**
 * Menyimpan konfigurasi GitHub yang dipakai alur "Build Otomatis":
 * repo tujuan, branch, nama file workflow, dan Personal Access Token (PAT).
 *
 * CATATAN KEAMANAN: token disimpan di SharedPreferences biasa (bukan
 * terenkripsi) langsung di penyimpanan aplikasi ini. Ini AMAN dari aplikasi
 * lain selama HP tidak di-root, tapi TIDAK terenkripsi kalau HP dicuri/hilang
 * dan berhasil di-root. Sangat disarankan:
 *   - Pakai "Fine-grained personal access token" (bukan classic token),
 *     dibatasi HANYA ke satu repo ini saja.
 *   - Beri izin token itu HANYA "Contents: Read & write" dan
 *     "Actions: Read & write" — jangan beri izin lain.
 *   - Set masa berlaku token (expiration), jangan pilih "No expiration".
 *   - Cabut (revoke) token dari GitHub kalau HP hilang.
 */
public class GitHubSettings {

    private static final String PREF_NAME = "rataapk_github_settings";
    private static final String KEY_OWNER = "owner";
    private static final String KEY_REPO = "repo";
    private static final String KEY_TOKEN = "token";
    private static final String KEY_BRANCH = "branch";
    private static final String KEY_WORKFLOW_FILE = "workflow_file";
    private static final String KEY_APP_LABEL = "app_label";
    private static final String KEY_PACKAGE_NAME = "package_name";

    private final SharedPreferences prefs;

    public GitHubSettings(Context context) {
        prefs = context.getApplicationContext().getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
    }

    public String getOwner() { return prefs.getString(KEY_OWNER, ""); }
    public String getRepo() { return prefs.getString(KEY_REPO, ""); }
    public String getToken() { return prefs.getString(KEY_TOKEN, ""); }
    public String getBranch() {
        String b = prefs.getString(KEY_BRANCH, "");
        return TextUtils.isEmpty(b) ? "main" : b;
    }
    public String getWorkflowFile() {
        String w = prefs.getString(KEY_WORKFLOW_FILE, "");
        return TextUtils.isEmpty(w) ? "build-apps.yml" : w;
    }
    public String getAppLabel() {
        String a = prefs.getString(KEY_APP_LABEL, "");
        return TextUtils.isEmpty(a) ? "RataAPK" : a;
    }
    public String getPackageName() {
        String p = prefs.getString(KEY_PACKAGE_NAME, "");
        return TextUtils.isEmpty(p) ? "id.rataapk.template" : p;
    }

    /** true kalau owner, repo, dan token sudah diisi (syarat minimum build otomatis). */
    public boolean sudahDikonfigurasi() {
        return !TextUtils.isEmpty(getOwner()) && !TextUtils.isEmpty(getRepo()) && !TextUtils.isEmpty(getToken());
    }

    public void simpan(String owner, String repo, String token, String branch,
                        String workflowFile, String appLabel, String packageName) {
        prefs.edit()
                .putString(KEY_OWNER, owner == null ? "" : owner.trim())
                .putString(KEY_REPO, repo == null ? "" : repo.trim())
                .putString(KEY_TOKEN, token == null ? "" : token.trim())
                .putString(KEY_BRANCH, branch == null ? "" : branch.trim())
                .putString(KEY_WORKFLOW_FILE, workflowFile == null ? "" : workflowFile.trim())
                .putString(KEY_APP_LABEL, appLabel == null ? "" : appLabel.trim())
                .putString(KEY_PACKAGE_NAME, packageName == null ? "" : packageName.trim())
                .apply();
    }

    public void hapusToken() {
        prefs.edit().remove(KEY_TOKEN).apply();
    }
}

File: ios/RataAPK/Sources/GitHubSettings.swift
import Foundation

/// Menyimpan konfigurasi GitHub yang dipakai alur "Build Otomatis" di iOS:
/// repo tujuan, branch, nama file workflow, dan Personal Access Token (PAT).
/// Mirror dari GitHubSettings.java di versi Android — perilaku dan nama
/// method disamakan supaya mudah dirawat berdampingan.
///
/// CATATAN KEAMANAN: sama seperti versi Android, token disimpan di
/// UserDefaults biasa (bukan Keychain, bukan terenkripsi). Aman dari
/// aplikasi lain di sandbox iOS normal, tapi tetap disarankan pakai
/// Fine-grained PAT yang dibatasi ke satu repo saja, dengan izin minimal
/// dan masa berlaku (expiration) yang jelas.
enum GitHubSettings {

    private static let kOwner = "rataapk_gh_owner"
    private static let kRepo = "rataapk_gh_repo"
    private static let kToken = "rataapk_gh_token"
    private static let kBranch = "rataapk_gh_branch"
    private static let kWorkflowFile = "rataapk_gh_workflow_file"
    private static let kAppLabel = "rataapk_gh_app_label"
    private static let kPackageName = "rataapk_gh_package_name"

    static var owner: String {
        get { UserDefaults.standard.string(forKey: kOwner) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: kOwner) }
    }

    static var repo: String {
        get { UserDefaults.standard.string(forKey: kRepo) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: kRepo) }
    }

    static var token: String {
        get { UserDefaults.standard.string(forKey: kToken) ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: kToken) }
    }

    static var branch: String {
        get {
            let b = UserDefaults.standard.string(forKey: kBranch) ?? ""
            return b.isEmpty ? "main" : b
        }
        set { UserDefaults.standard.set(newValue, forKey: kBranch) }
    }

    static var workflowFile: String {
        get {
            let w = UserDefaults.standard.string(forKey: kWorkflowFile) ?? ""
            return w.isEmpty ? "build-apps.yml" : w
        }
        set { UserDefaults.standard.set(newValue, forKey: kWorkflowFile) }
    }

    static var appLabel: String {
        get {
            let a = UserDefaults.standard.string(forKey: kAppLabel) ?? ""
            return a.isEmpty ? "RataAPK" : a
        }
        set { UserDefaults.standard.set(newValue, forKey: kAppLabel) }
    }

    static var packageName: String {
        get {
            let p = UserDefaults.standard.string(forKey: kPackageName) ?? ""
            return p.isEmpty ? "id.rataapk.template" : p
        }
        set { UserDefaults.standard.set(newValue, forKey: kPackageName) }
    }

    /// true kalau owner, repo, dan token sudah diisi (syarat minimum build otomatis).
    static func sudahDikonfigurasi() -> Bool {
        return !owner.isEmpty && !repo.isEmpty && !token.isEmpty
    }

    static func simpan(owner: String, repo: String, token: String, branch: String,
                        workflowFile: String, appLabel: String, packageName: String) {
        self.owner = owner.trimmingCharacters(in: .whitespacesAndNewlines)
        self.repo = repo.trimmingCharacters(in: .whitespacesAndNewlines)
        self.token = token.trimmingCharacters(in: .whitespacesAndNewlines)
        self.branch = branch.trimmingCharacters(in: .whitespacesAndNewlines)
        self.workflowFile = workflowFile.trimmingCharacters(in: .whitespacesAndNewlines)
        self.appLabel = appLabel.trimmingCharacters(in: .whitespacesAndNewlines)
        self.packageName = packageName.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func hapusToken() {
        UserDefaults.standard.removeObject(forKey: kToken)
    }
}
